controller/api/destination/profile_translator.go

```diff
@@ -0,0 +1,323 @@
+	"github.com/linkerd/linkerd2/pkg/util"
+	logging "github.com/sirupsen/logrus"
+)
+
+var (
+	defaultRetryBudget = pb.RetryBudget{
+		MinRetriesPerSecond: 10,
+		RetryRatio:          0.2,
+		Ttl: &duration.Duration{
+			Seconds: 10,
+		},
+	}
+
+	defaultServiceProfile = pb.DestinationProfile{
+		Routes:      []*pb.Route{},
+		RetryBudget: &defaultRetryBudget,
+	}
+
+	defaultRouteTimeout = 10 * time.Second
+)
+
+// implements the ProfileUpdateListener interface
+type profileTranslator struct {
+	stream pb.Destination_GetProfileServer
+	log    *logging.Entry
+}
+
+func newProfileTranslator(stream pb.Destination_GetProfileServer, log *logging.Entry) *profileTranslator {
+	return &profileTranslator{
+		stream: stream,
+		log:    log.WithField("component", "profile-translator"),
+	}
+}
+
+func (pt *profileTranslator) Update(profile *sp.ServiceProfile) {
+	if profile == nil {
+		pt.stream.Send(&defaultServiceProfile)
+		return
+	}
+	destinationProfile, err := toServiceProfile(profile)
+	if err != nil {
+		pt.log.Error(err)
+		return
+	}
+	pt.log.Debugf("Sending profile update: %+v", destinationProfile)
+	pt.stream.Send(destinationProfile)
+}
+
+func toDuration(d time.Duration) *duration.Duration {
+	return &duration.Duration{
+		Seconds: int64(d / time.Second),
+		Nanos:   int32(d % time.Second),
+	}
+}
+
+// toServiceProfile returns a Proxy API DestinationProfile, given a
+// ServiceProfile.
+func toServiceProfile(profile *sp.ServiceProfile) (*pb.DestinationProfile, error) {
+	routes := make([]*pb.Route, 0)
+	for _, route := range profile.Spec.Routes {
+		pbRoute, err := toRoute(profile, route)
+		if err != nil {
+			return nil, err
+		}
+		routes = append(routes, pbRoute)
+	}
+	budget := defaultRetryBudget
+	if profile.Spec.RetryBudget != nil {
+		budget.MinRetriesPerSecond = profile.Spec.RetryBudget.MinRetriesPerSecond
+		budget.RetryRatio = profile.Spec.RetryBudget.RetryRatio
+		ttl, err := time.ParseDuration(profile.Spec.RetryBudget.TTL)
+		if err != nil {
+			return nil, err
+		}
+		budget.Ttl = toDuration(ttl)
+	}
+	return &pb.DestinationProfile{
+		Routes:      routes,
+		RetryBudget: &budget,
+	}, nil
+}
+
+// toRoute returns a Proxy API Route, given a ServiceProfile Route.
+func toRoute(profile *sp.ServiceProfile, route *sp.RouteSpec) (*pb.Route, error) {
+	cond, err := toRequestMatch(route.Condition)
+	if err != nil {
+		return nil, err
+	}
+	rcs := make([]*pb.ResponseClass, 0)
+	for _, rc := range route.ResponseClasses {
+		pbRc, err := toResponseClass(rc)
+		if err != nil {
+			return nil, err
+		}
+		rcs = append(rcs, pbRc)
+	}
+	timeout := defaultRouteTimeout
+	if route.Timeout != "" {
+		timeout, err = time.ParseDuration(route.Timeout)
+		if err != nil {
+			logging.Errorf(
+				"failed to parse duration for route '%s' in service profile '%s' in namespace '%s': %s",
+				route.Name,
+				profile.Name,
+				profile.Namespace,
+				err,
+			)
+			timeout = defaultRouteTimeout
+		}
+	}
+	ret := pb.Route{
```

Comment:
instead of creating `ret` could you return a `&pb.Route` directly as you did in line 87?

```
	return &pb.Route{
		Condition:       cond,
		ResponseClasses: rcs,
		MetricsLabels:   map[string]string{"route": route.Name},
		IsRetryable:     route.IsRetryable,
		Timeout:         toDuration(timeout),
	}, nil
```

Reply:
nice catch, thanks